snuba/query/parser/validation/functions.py

```diff
@@ -30,28 +34,38 @@ class FunctionCallsValidator(ExpressionValidator):
     a default one applied to all queries and one a mapping per dataset.
     """
 
-    def validate(self, exp: Expression, entity: Entity) -> None:
+    def validate(self, exp: Expression, data_source: DataSource) -> None:
         if not isinstance(exp, FunctionCall):
             return
 
-        entity_validators = entity.get_function_call_validators()
-        common_function_validators = (
-            entity_validators.keys() & default_validators.keys()
-        )
-        if common_function_validators:
-            logger.warning(
-                "Dataset validators are overlapping with default ones. Entity: %s. Overlap %r",
-                entity,
-                common_function_validators,
-                exc_info=True,
+        entity_validators: Mapping[str, FunctionCallValidator] = {}
+        data_model: Optional[ColumnSet] = None
+        if isinstance(data_source, QueryEntity):
+            entity = get_entity(data_source.key)
+            entity_validators = entity.get_function_call_validators()
+
+            common_function_validators = (
+                entity_validators.keys() & default_validators.keys()
             )
+            if common_function_validators:
+                logger.warning(
+                    "Dataset validators are overlapping with default ones. Entity: %s. Overlap %r",
+                    entity,
+                    common_function_validators,
+                    exc_info=True,
+                )
+            data_model = entity.get_data_model()
+        elif isinstance(data_source, (Query, JoinClause)):
+            data_model = data_source.get_columns()
+        else:
+            return
 
         validators = ChainMap(default_validators, entity_validators)
```

Comment:
Is this not skipping all the entity validators on joins? Do you have a plan for them ?

Reply:
Currently we have no entities using the function call validators at all, so my plan was actually to deprecate all of this. It seems like we could fold the small amount of validation into a query processor or into the parser.

Reply:
I don't think that will be enough. Doing this type of entity specific validation as a processor will mean doing it only on the subqueries after the push down. Which means not validating the expressions in the parent query according to the types of the columns of the subqueries. The columns in the columnsets exposed by a subquery are all of type `Any`.
I think we either validate the join before the push down or we propagate up the types and validate both the subqueries and the parent query.

The amount of validation here is small because we never went ahead but it is supposed to become much more important specifically by limiting the available functions we expose.

So if there are no entity specific validators now, I think it would be ok to skip that phase. But I would not deprecate validation at this stage as it would be easier to validate entity specific expression before push down.

Reply:
We definitely don't have any now. I'll add a TODO and backlog task to figure out entity validation in joins.